### PR TITLE
Application 22 55 tt

### PR DESCRIPTION
**Pull request by tylertomlinson**
User stories 22-25 
Added a joins table with pet id and application id linking with many to many relationship 
Application process complete 
User can create application for multiple pets. Application can be approved for multiple pet
Application can be approved and revoked 
No more than one application approved per pet (other applications stay on file)

Model test covering 100% 
Feature test covering 95% (need to update test only. no added tested needed)

Paying down technical debt:
Made helper methods that live on the `ApplicationController`
Implemented a lot of string interpolation and some css styling

### CODE REVIEW

File: app/controllers/applications_controller.rb [hunk 1]
```diff
@@ -6,6 +6,32 @@ def index
...
+    if pet.status == "pending"
+      flash[:notice] = "Not Taking New Applications At This Time"
+      redirect_to "/applications/#{application.id}"
+    else
+      pet.update_attribute(:status, "pending")
+      pet_application = PetApplication.where(application_id: params[:application_id], pet_id: params[:pet_id]).first
+      pet_application.update(approved: true)
+
+      redirect_to "/pets/#{pet.id}"
+    end
```
Comment (StephanieFriend) [inline app/controllers/applications_controller.rb:24]: Looks great! May want to look into splitting up if we have time.

File: app/controllers/pets_controller.rb [hunk 2]
```diff
@@ -16,8 +16,12 @@ def create
 
   def show
     @pet = get_pet_info
+    if @pet.pet_applications.find_by(approved: true) != nil
+     @application_name = @pet.pet_applications.find_by(approved: true).application.name
+     end
```
Comment (StephanieFriend) [inline app/controllers/pets_controller.rb:21]: Love this logic

File: app/views/application/show.html.erb [hunk 3]
```diff
@@ -17,11 +17,20 @@
...
+    </section>
+    <section>
+      <h3>Pets Applied For</h3>
+      <% @application.pets.each do |pet|%>
+      <section id="pet-links<%= pet.id %>">
+        <%= link_to pet.name, "/pets/#{pet.id}" %><br/>
+        <% @petapps.each do |petapp| %>
+        <% if petapp.approved && petapp.application_id == @application.id%>
+          <%= link_to "Revoke Application For: #{pet.name}", "/pets/#{pet.id}/applications/#{@application.id}/revoke_status", method: :patch %>
+        <% else %>
```
Comment (StephanieFriend) [inline app/views/application/show.html.erb:29]: If time permits we may want to look into moving this logic into the controller

File: config/routes.rb [hunk 4]
```diff
@@ -36,4 +36,9 @@
   post '/applications', to: 'applications#create'
   get '/applications/:id', to: 'applications#show'
   get '/pets/:id/pet_applications', to: 'applications#pet_application'
+
+  patch '/pets/:pet_id/applications/:application_id/approve_status', to: 'applications#approve_status'
+  patch '/pets/:pet_id/applications/:application_id/revoke_status', to: 'applications#revoke_status'
+
+  # patch '/pets/:pet_id/applications/:application_id/:new_status', to:'pets#update_status'
```
Comment (StephanieFriend) [inline config/routes.rb:43]: Have to remember to remove comments

File: spec/features/applications/apply_for_multiple_pets_spec.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+#     # expect(page).to have_content "On hold for #{@application_1.name}"
+#
+#     visit "/applications/#{@application_1.id}"
+#
+#     click_link "Approve application for: #{@pet_2.name}"
+#     expect(current_path).to eq "/pets/#{@pet_2.id}"
+#
+#     expect(page).to have_content "Adoptable: pending"
+#     # expect(page).to have_content "On hold for #{@application_1.name}"
+#     end
```
Comment (StephanieFriend) [inline spec/features/applications/apply_for_multiple_pets_spec.rb:36]: Why is a test commented out?

File: spec/features/applications/approving_application_spec.rb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+#     expect(current_path).to eq "/pets/#{@pet_2.id}"
+#
+#     expect(page).to have_content "Adoptable: pending"
+#     expect(page).to have_content "On hold for #{@application_1.name}"
+#
+#     visit "/applications/#{@application_2.id}"
+#     expect(page).not_to have_link("Approve application for: #{@pet_2.name}")
+#     expect(page).to have_link("Approve application for: #{@pet_1.name}")
+#
+#     visit "/pets/#{@pet_2.id}/applications"
```
Comment (StephanieFriend) [inline spec/features/applications/approving_application_spec.rb:62]: This one too?

File: spec/features/pets/user_can_see_pets_with_info_spec.rb [hunk 7]
```diff
@@ -34,6 +33,5 @@
     expect(page).to_not have_content(@pet_1.description)
     expect(page).to_not have_content(@pet_1.age)
     expect(page).to_not have_content(@pet_1.sex)
```
Comment (StephanieFriend) [inline spec/features/pets/user_can_see_pets_with_info_spec.rb:35]: Excited that tests are looking great!

**Review by StephanieFriend — APPROVED** [review]
This was done by both of us. We passed back and forth our branches until it was working and felt comfortable pushing into master

File: spec/features/applications/apply_for_multiple_pets_spec.rb (hunk 5 above)
Comment (tylertomlinson) [reply spec/features/applications/apply_for_multiple_pets_spec.rb:36]: Need to be able to get the routes correct

**Pull request merged**